test-data/unit/check-functions.test

```diff
@@ -2067,3 +2067,26 @@ def some_method(self: badtype): pass # E: Name 'badtype' is not defined
 def fn(
         a: badtype) -> None: # E: Name 'badtype' is not defined
     pass
+
+[case testInferredTypeMismatchReturnType]
+from typing import Union, Dict, List
+def f() -> List[Union[int, str]]:
+    x = ['a']
+    return x # E: Incompatible return value type (got List[str], expected List[Union[int, str]]) \
+# N: Perhaps you need a type annotation for "x"? Suggestion: List[Union[int, str]]
+
+def g() -> Dict[str, Union[int, str]]:
+    x = {'a': 'a'}
+    return x # E: Incompatible return value type (got Dict[str, str], expected Dict[str, Union[int, str]]) \
+# N: Perhaps you need a type annotation for "x"? Suggestion: Dict[str, Union[int, str]]
+
+def h() -> List[Union[int, float]]:
+    x = ['a']
+    return x # E: Incompatible return value type (got List[str], expected List[Union[int, float]])
+
+def i() -> List[Union[int, float]]:
+    x = [1] # Type: List[int]
+    return x # E: Incompatible return value type (got List[int], expected List[Union[int, float]]) \
+# N: Perhaps you need a type annotation for "x"? Suggestion: List[Union[int, float]]
```

Comment:
Almost there! I'd suggest adding test cases that cover cases that are mostly like these but that don't actually trigger the new message (i.e. one of the conditions in `make_inferred_type_note` is false).